### PR TITLE
`FunctionNode` automatic fallback of array attributes in forward

### PR DESCRIPTION
**Pull request by niboshi**
In ChainerX FunctionNode fallback, type mismatch will happen if the function assigns an array as its attribute.
This PR implements automatic conversion of such attributes.

### CODE REVIEW

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
`test_gaussian.py` seems to be failing due to device mismatch. Could you fix it?

**Comment by hvy** [comment]
I assume PReLU and Gaussian are just two examples but that there are many other functions that can make use of this fix. I personally think it's worth considering a separate interface in the future for assigning/retrieving ndarrays inside FunctionNodes, rather than directly calling __{set,get}attr__.

**Comment by niboshi** [comment]
I guess test failures should be fixed by #5777.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
Could you check the ChainerX CI failures?

**Comment by niboshi** [comment]
Rebased and fixed.

**Comment by hvy** [comment]
jenkins, test this please.

File: chainer/function_node.py [hunk 1]
```diff
@@ -398,23 +411,64 @@ def check_type_forward(self, in_types):
...
-            in_data.append(x_data)
+            in_data.append(fallback_data)
 
         in_data = tuple(in_data)
-        return chainerx_in_data, in_data
+        return chainerx_in_data, in_data, device
+
+    def _make_chainerx_forward_fallback_class(self, device):
+        # Creates a fabricated class based on the concerete FunctionNode class,
+        # equipped with the automatic attribute fallback.
```
Comment (hvy) [inline chainer/function_node.py]: Do you think you could make the description a bit more verbose? I think difficult to understand for new developers.

**Comment by hvy** [comment]
Sorry just added a single last comment.

**Comment by niboshi** [comment]
PTAL

File: chainer/function_node.py [hunk 2]
```diff
@@ -398,23 +411,79 @@ def check_type_forward(self, in_types):
...
         in_data = tuple(in_data)
-        return chainerx_in_data, in_data
+        return chainerx_in_data, in_data, device
+
+    def _make_chainerx_forward_fallback_class(self, device):
+        # Creates a fabricated class based on the concerete FunctionNode class,
+        # equipped with the automatic attribute fallback, which is enabled
+        # during the forward function.
+        #
+        # In the fallback mechanism, wnen an array with the fallback ndarray
```
Comment (hvy) [inline chainer/function_node.py]:
```suggestion
        # In the fallback mechanism, when an array with the fallback ndarray
```

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
LGTM. The AppVeyor failure seems unrelated.

**Pull request merged**